### PR TITLE
GeneralFeedback

### PR DESCRIPTION
**Pull request by norbertK**
implements BAY-942

### CODE REVIEW

File: src/modules/feedback/components/rating/FiveButtonRating.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * @enum {string}
+ */
+export const Rating = Object.freeze({
+	NONE: '0',
+	TERRIBLE: '1',
+	BAD: '2',
+	SATISFIED: '3',
+	GOOD: '4',
+	EXCELLENT: '5'
+});
```
Comment (taulinger) [inline src/modules/feedback/components/rating/FiveButtonRating.js:22]: Let's convert the values to numbers (0-5) and add a test for this enum

File: test/modules/feedback/components/rating/FiveButtonRating.test.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+			// act
+			const starButtons = element.shadowRoot.querySelectorAll('.star-button');
+			starButtons.forEach((starButton) => {
+				starButton.click();
+			});
+
+			// assert
+			expect(ratingSpy).toHaveBeenCalledTimes(5);
+		});
+	});
```
Comment (taulinger) [inline test/modules/feedback/components/rating/FiveButtonRating.test.js:84]: Replace both tests with a new describe block and test:
```
describe('when rating is set (via property)', () => {
    it('updates the UI', async () => {
.
.
.
  it('calls a registered "onChange" handler', async () => {

```

File: src/modules/feedback/components/rating/FiveButtonRating.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+const Update_Rating = 'update_rating';
+
+export class FiveButtonRating extends MvuElement {
```
Comment (taulinger) [inline src/modules/feedback/components/rating/FiveButtonRating.js:26]: `StarsRatingPanel`

File: src/modules/feedback/components/rating/FiveButtonRating.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			})
+		);
+	}
+
+	get rating() {
+		return this.getModel().rating;
+	}
+
+	static get tag() {
+		return 'ba-mvu-fivebuttonrating';
```
Comment (taulinger) [inline src/modules/feedback/components/rating/FiveButtonRating.js:111]: `ba-stars-rating-panel`

File: src/modules/feedback/components/rating/FiveButtonRating.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	_onRatingClick(rating) {
+		this.rating = rating;
+	}
+
+	set rating(value) {
+		this.signal(Update_Rating, value);
+		this.dispatchEvent(
+			new CustomEvent('rating', {
```
Comment (taulinger) [inline src/modules/feedback/components/rating/FiveButtonRating.js:100]: Let's name the event `change`

File: src/modules/feedback/components/rating/FiveButtonRating.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		`;
+	}
+
+	_onRatingClick(rating) {
+		this.rating = rating;
+	}
+
+	set rating(value) {
+		this.signal(Update_Rating, value);
+		this.dispatchEvent(
```
Comment (taulinger) [inline src/modules/feedback/components/rating/FiveButtonRating.js:99]: Here we would fire the event every time the setter is called even if the model hasn't changed.
We could improve this by registering an observer for our model (in onInitialize()) and firing the event when our observer was called.

File: test/modules/feedback/components/rating/FiveButtonRating.test.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+			expect(starButtons.length).toBe(5);
+			starButtons.forEach((starButton) => {
+				starButton.click();
+			});
+
+			// assert
+			expect(onRatingClickSpy).toHaveBeenCalled();
+			expect(onRatingClickSpy).toHaveBeenCalledTimes(5);
+		});
+	});
```
Comment (taulinger) [inline test/modules/feedback/components/rating/FiveButtonRating.test.js:103]: it('updates the UI', async () => {
.
.
.
  it('calls a registered "onChange" handler', async () => {

File: test/modules/feedback/components/rating/StarsRatingPanel.test.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+			const starButtons = element.shadowRoot.querySelectorAll('.star-button');
+			starButtons.forEach((starButton) => {
+				starButton.click();
+			});
+
+			// assert
+			expect(starButtons.length).toBe(5);
+			expect(onRatingClickSpy).toHaveBeenCalled();
+			expect(onRatingClickSpy).toHaveBeenCalledTimes(5);
+		});
```
Comment (taulinger) [inline test/modules/feedback/components/rating/StarsRatingPanel.test.js:102]: we also need a test `updates the UI` -> some of the stars should be displayed filled now

File: test/modules/feedback/components/rating/StarsRatingPanel.test.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+			// assert
+			expect(ratingSpy).toHaveBeenCalled();
+			expect(element.rating).toBe(Rating.BAD);
+
+			const starButtons = element.shadowRoot.querySelectorAll('.star-button');
+			expect(starButtons.length).toBe(5);
+			starButtons.forEach((starButton) => {
+				expect(starButton.onClick).toEqual(element.onRatingClick);
+				expect(starButton.title).toMatch(new RegExp(`^${prefix}`));
```
Comment (taulinger) [inline test/modules/feedback/components/rating/StarsRatingPanel.test.js:81]: does not check if the starts are displayed filled now

File: src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+import { BA_FORM_ELEMENT_VISITED_CLASS } from '../../../../utils/markup';
+import { GeneralFeedback } from '../../../../services/FeedbackService';
+
+const Update_Rating = 'update_rating';
+const Update_Description = 'update_description';
+const Update_EMail = 'update_email';
+
+/**
+ * Contains a form for submitting a general feedback.
+ * @class
```
Comment (taulinger) [inline src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js:19]: `* @property {Function} onSubmit` is missing

File: test/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.test.js [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+			descriptionInput.value = 'description';
+			descriptionInput.dispatchEvent(new Event('input'));
+
+			const submitButton = element.shadowRoot.querySelector('#button0');
+
+			// act
+			submitButton.click();
+
+			// assert
+			expect(saveGeneralFeedbackSpy).toHaveBeenCalled();
```
Comment (taulinger) [inline test/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.test.js:222]: Unneeded check (see next line)

File: src/modules/feedback/i18n/feedback.provider.js [hunk 12]
```diff
@@ -2,57 +2,77 @@ export const provide = (lang) => {
 	switch (lang) {
 		case 'en':
 			return {
-				feedback_mapFeedback_header: 'Send Feedback',
-				feedback_mapFeedback_could_not_save: 'Your feedback could not be saved',
-				feedback_mapFeedback_saved_successfully: 'Your feedback was successfully submitted',
-				feedback_mapFeedback_categorySelection: 'Category',
+				feedback_generalFeedback_changeDescription_error: 'Required field',
+				feedback_generalFeedback_changeDescription_helper: 'Required field',
```
Comment (taulinger) [inline src/modules/feedback/i18n/feedback.provider.js:6]: Let's replace all individual error keys with one valid for all feedback components,
e.g:
```
feedback_required_field: 'Required field'
feedback_invalid_eMail: 'No valid e-mail address
```

File: src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js (hunk 10 above)
Comment (taulinger) [reply src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js:19]: Also in `MapFeedbackPanel.js`

File: src/modules/feedback/components/rating/StarsRatingPanel.js [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	GOOD: 4,
+	EXCELLENT: 5
+});
+
+const Update_Rating = 'update_rating';
+
+/**
+ * Rating component
+ *
+ * Events;
```
Comment (taulinger) [inline src/modules/feedback/components/rating/StarsRatingPanel.js:29]: Uneeded

File: src/modules/feedback/i18n/feedback.provider.js [hunk 14]
```diff
@@ -2,57 +2,59 @@ export const provide = (lang) => {
...
 		case 'en':
 			return {
-				feedback_mapFeedback_header: 'Send Feedback',
-				feedback_mapFeedback_could_not_save: 'Your feedback could not be saved',
-				feedback_mapFeedback_saved_successfully: 'Your feedback was successfully submitted',
-				feedback_mapFeedback_categorySelection: 'Category',
+				feedback_changeDescription: 'Your message',
+				feedback_eMail_error: 'No valid e-mail address',
+				feedback_eMail_helper: 'Optional',
+				feedback_eMail: 'Your email address',
```
Comment (taulinger) [inline src/modules/feedback/i18n/feedback.provider.js:8]: Let's move all common feedback messages to the top

File: src/modules/feedback/i18n/feedback.provider.js [hunk 15]
```diff
@@ -2,57 +2,59 @@ export const provide = (lang) => {
...
-				feedback_toggleFeedback_generalButton_sub: 'Make suggestions for changes or improvements, or report errors.'
+				feedback_mapFeedback_privacyPolicy: 'Privacy Policy',
+				feedback_mapFeedback_text_after: ' and fill in the fields. When the change is added to BayernAtlas, it will appear publicly.',
+				feedback_mapFeedback_text_before: 'Please draw your change on the',
+				feedback_mapFeedback_text_map: 'Map',
+				feedback_mapFeedback: 'Map - Feedback',
+				feedback_required_field_error: 'Required field',
+				feedback_required_field_helper: 'Required field',
+				feedback_saved_successfully: 'Your feedback was successfully submitted',
+				feedback_submit: 'Submit',
```
Comment (taulinger) [inline src/modules/feedback/i18n/feedback.provider.js:26]: Let's move all common feedback messages to the top

**Review by taulinger — COMMENTED** [review]
Can we remove all double expectations in both `GeneralFeedbackPanel.test.js`  and `StarsRaingPanel.test.js` please?

Example:
```
expect(onRatingClickSpy).toHaveBeenCalled();
expect(onRatingClickSpy).toHaveBeenCalledTimes(5);
```
The second expectation  makes the first one obsolete

File: src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+			<div class="ba-form-element">
+				<input type="email" id="email" placeholder="${translate('feedback_eMail')}" .value="${generalFeedback.email}" @input="${handleEmailChange}" />
+				<label for="email" class="control-label">${translate('feedback_eMail')}</label>
+				<i class="bar"></i>
+				<i class="icon error"></i>
+				<label class="helper-label">${translate('feedback_eMail_helper')}</label>
+				<label class="error-label">${translate('feedback_eMail_error')}</label>
+			</div>
+
+			<p id="feedback_mapFeedback_disclaimer" class="map-feedback__disclaimer" id="mapFeedback_disclaimer">
```
Comment (thiloSchlemmer) [inline src/modules/feedback/components/generalFeedback/GeneralFeedbackPanel.js:139]: there should be only one id...

In GeneralFeedbackPanel.js should 'mapFeedback' not occur.

If the name is an Element-ID -> rename id to `generalFeedback*`.
If the name is an i18n-ID -> rename the id in ALL places : `feedback_mapFeedback_disclaimer` -> `feedback_disclaimer`

File: src/modules/feedback/components/generalFeedback/generalFeedbackPanel.css
Comment (thiloSchlemmer) [inline src/modules/feedback/components/generalFeedback/generalFeedbackPanel.css:1]: Clean up the file from unused css classes.
The subsequent styling step will add them back if necessary.

File: src/modules/feedback/components/mapFeedback/MapFeedbackPanel.js [hunk 17]
```diff
@@ -232,34 +232,28 @@ export class MapFeedbackPanel extends MvuElement {
...
+						<input type="email" id="email" .value="${mapFeedback.email}" @input="${onEmailChange}" placeholder="${translate('feedback_eMail')}" />
+						<label for="email" class="control-label">${translate('feedback_eMail')}</label>
 						<i class="bar"></i>
 						<i class="icon error"></i>
-						<label class="helper-label">${translate('feedback_mapFeedback_eMail_helper')}</label>
-						<label class="error-label">${translate('feedback_mapFeedback_eMail_error')}</label>
+						<label class="helper-label">${translate('feedback_eMail_helper')}</label>
+						<label class="error-label">${translate('feedback_eMail_error')}</label>
 					</div>
 					<p id="feedback_mapFeedback_disclaimer" class="map-feedback__disclaimer" id="mapFeedback_disclaimer">
```
Comment (thiloSchlemmer) [inline src/modules/feedback/components/mapFeedback/MapFeedbackPanel.js:251]: there should be only one id...

**Review by thiloSchlemmer — COMMENTED** [review]
See inline comments

**Pull request merged**